kong/plugins/prometheus/exporter.lua

```diff
@@ -168,6 +174,50 @@ local function collect()
                  "/metrics endpoint: ", err)
   end
 
+  -- upstream targets accessible?
+  local balancer = require "kong.runloop.balancer"
+  local upstreams_dict = balancer.get_all_upstreams()
+  for upstream_name, upstream_id in pairs(upstreams_dict) do
+    -- same logic from kong.db.dao.targets
+    local health_info
+    health_info, err = balancer.get_upstream_health(upstream_id)
+    if err then
+      ngx.log(ngx.ERR, "failed getting upstream health: ", err)
+    end
+
+    if health_info then
+      local status = 0 -- healthchecks is off
+      for target_name, target_info in pairs(health_info) do
+        if target_info ~= nil and
+          #target_info.addresses > 0 then
+          for _, address in ipairs(target_info.addresses) do
+            if address.health == "HEALTHY" then
+              status = 1
+              break
+            elseif address.health == "UNHEALTHY" then
+              status = 2
+            end
+          end
+        else
+          status = 3 -- dns error
+        end
+        if status == 1 then
+          metrics.upstream_target_health:set(1, { upstream_name, target_name, "healthy" })
```

Comment:
if we allocate a table with 3 array elements outside of for loop then it can be reused all the time. https://github.com/Kong/kong-plugin-prometheus/blob/master/kong/plugins/prometheus/exporter.lua#L97 can be a reference.

Reply:
nice!

Reply:
I only get the `target_name` in the inner loop. Using a table in this situation would bring any benefit?

Reply:
it will reduce some GC pressure for creating/recycling tables. basically `metrics.upstream_target_health:set(1, { upstream_name, target_name, "healthy" })` will create a new table.

Reply:
@fffonion let me know if the newest code helps